web/notes/notes.go

```diff
@@ -23,17 +24,42 @@ import (
 // CreateNote is the API handler for POST /notes. It creates a note, aka a file
 // with a set of metadata to enable collaborative edition.
 func CreateNote(c echo.Context) error {
-	if err := middlewares.AllowWholeType(c, permission.POST, consts.Files); err != nil {
-		return err
-	}
-
+	inst := middlewares.GetInstance(c)
 	doc := &note.Document{}
 	if _, err := jsonapi.Bind(c.Request().Body, doc); err != nil {
 		return err
 	}
 	doc.CreatedBy = getCreatedBy(c)
 
-	inst := middlewares.GetInstance(c)
+	// We first look if we have a permission on the whole doctype, as it is
+	// cheap. If not, we look on more finer permissions, which is a bit more
+	// complicated and costly, but is needed for creating a note in a shared by
+	// link folder for example.
+	if err := middlewares.AllowWholeType(c, permission.POST, consts.Files); err != nil {
+		dirID, errd := doc.GetDirID(inst)
+		if errd != nil {
+			return err
+		}
+		fileDoc, errf := vfs.NewFileDoc(
+			"tmp.cozy-note", // We don't care, but it can't be empty
```

Comment:
I may misunderstand what we do here, but if we create a temporary file to check permissions, shouldn't we delete it afterwards?

Reply:
The file is created in memory, but not persisted in CouchDB/swift. So there is no need to delete it, the Garbage Collector will clean the memory.